eppoclient/evalbandits.go

```diff
@@ -0,0 +1,234 @@
+			result.Numeric[key] = value
+		case string:
+			result.Categorical[key] = value
+		case bool:
+			result.Categorical[key] = strconv.FormatBool(value)
+		}
+	}
+	return result
+}
+
+func (self ContextAttributes) toGenericAttributes() Attributes {
+	result := make(Attributes)
+	for key, value := range self.Numeric {
+		result[key] = value
+	}
+	for key, value := range self.Categorical {
+		result[key] = value
+	}
+	return result
+}
+
+type banditEvaluationContext struct {
+	flagKey           string
+	subjectKey        string
+	subjectAttributes ContextAttributes
+	actions           map[string]ContextAttributes
+}
+
+type banditEvaluationDetails struct {
+	flagKey           string
+	subjectKey        string
+	subjectAttributes ContextAttributes
+	result            *banditEvaluationResult
+}
+type banditEvaluationResult struct {
+	actionKey        string
+	actionAttributes ContextAttributes
+	actionScore      float64
+	actionWeight     float64
+	gamma            float64
+	optimalityGap    float64
+}
+
+type action struct {
+	key        string
+	attributes ContextAttributes
+}
+
+func (model *banditModelData) evaluate(ctx banditEvaluationContext) banditEvaluationDetails {
+	// THere's currently no way to change totalShards in bandit evaluation.
+	var totalShards int64 = 10_000
+
+	nActions := len(ctx.actions)
+	if nActions == 0 {
+		return banditEvaluationDetails{
+			flagKey:           ctx.flagKey,
+			subjectKey:        ctx.subjectKey,
+			subjectAttributes: ctx.subjectAttributes,
+			result:            nil,
+		}
+	}
+
+	bestAction, bestScore := "", math.Inf(-1)
+	scores := make(map[string]float64)
+	for actionKey, actionAttributes := range ctx.actions {
+		score := model.scoreAction(ctx.subjectAttributes, action{key: actionKey, attributes: actionAttributes})
+		scores[actionKey] = score
+		if score > bestScore {
+			bestAction, bestScore = actionKey, score
+		}
+	}
+
+	// adjust probability floor for number of actions to control the sum
+	minProbability := model.ActionProbabilityFloor / float64(nActions)
+
+	weights := make(map[string]float64)
+	totalWeight := 0.0
+	for actionKey, score := range scores {
+		if actionKey != bestAction { // best action is assigned the remainder weight
+			weight := math.Max(minProbability, 1.0/(float64(nActions)+model.Gamma*(bestScore-score)))
+			weights[actionKey] = weight
+			totalWeight += weight
+		}
+	}
+
+	remainderWeight := math.Max(0.0, 1.0-totalWeight)
+	weights[bestAction] = remainderWeight
+
+	actions := make([]string, nActions)
+	shards := make(map[string]int64)
+	{
+		i := 0
+		for actionKey := range ctx.actions {
+			actions[i] = actionKey
+			shards[actionKey] = getShard(ctx.flagKey+"-"+ctx.subjectKey+"-"+actionKey, totalShards)
+			i++
+		}
+	}
+
+	sort.Slice(actions, func(i, j int) bool {
+		a1 := actions[i]
+		a2 := actions[j]
+		v1 := shards[a1]
+		v2 := shards[a2]
+		if v1 < v2 {
+			return true
+		} else if v1 > v2 {
+			return false
+		} else {
+			// tie-breaking
+			return a1 < a2
```

Comment:
I think comments could go a long way here. We're "shuffling" the actions by sorting by a randomly assigned shard number (which--while unlikely--two actions _may_ share), and then tie breaking with action names, which two actions will not share.